lib/irb/cmd/history.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+require "stringio"
+require_relative "nop"
+require_relative "../pager"
+
+module IRB
+  # :stopdoc:
+
+  module ExtendCommand
+    class History < Nop
+      category "IRB"
+      description "Shows the input history. `-g [query]` or `-G [query]` allows you to filter the output."
+
+      def self.transform_args(args)
+        if match = args&.match(/\A(?<args>.+\s|)(-g|-G)\s+(?<grep>[^\s]+)\s*\n\z/)
+          args = match[:args]
+          "#{args}#{',' unless args.chomp.empty?} grep: /#{match[:grep]}/"
+        else
+          args
+        end
+      end
+
+      def execute(*arg, grep: nil)
+        formatted_inputs = irb_context.io.class::HISTORY.each_with_index.reverse_each.map do |input, index|
+          header = "#{index}: "
+
+          first_line, *other_lines = input.split("\n") || [""]
+          first_line.prepend header
+
+          truncated_lines = other_lines.slice!(1..) # Show 1 additional line (2 total)
+          other_lines << "..." if truncated_lines&.any?
+
+          other_lines.map! do |line|
+            " " * header.length + line
+          end
+
+          StringIO.new.tap { |io| io.puts(first_line, *other_lines) }.string
```

Comment:
This line's a bit ugly. There might be a better write to rewrite it while addressing your comment about `puts a, *b` vs `puts [a, *b].join("\n")`.

In https://github.com/ruby/irb/pull/761/commits/61dc0e54a1179f97952fa3ced66ca52ce4a7035d I rewrote it as `output.puts first_line, *truncated_input`. However, after adding the `-g` option, I lost the need for the `output` StringIO.

Reply:
Yeah, it would be better if  it can be simply done without StringIO.

Reply:
What are your thoughts on this?

```ruby
formatted = [first_line, *other_lines].join("\n")
formatted += "\n" unless formatted.end_with?("\n")
formatted
```

It's not the cleanest, but should hopefully mimic `puts`'s new line formatting.
> Writes a newline after each that does not already end with a newline sequence.
https://docs.ruby-lang.org/en/master/ARGF.html#method-i-puts

Is there a better way to write this?

Reply:
Here you go! https://github.com/ruby/irb/pull/761/commits/f2c3d9af8fbb859c40a71dae5f0477f03214c23d